pyquil/quil.py

```diff
@@ -164,6 +165,46 @@ def if_then(self, classical_reg, if_program, else_program=None):
         branch.Else.inst(else_program)
         return self.inst(branch)
 
+    def dagger(self):
+        """
+        Creates the conjugate transpose of the Quil program. The program must not
+        contain any irreversible actions (measurement, control flow, qubit allocation).
+
+        :return: The Quil program's inverse
+        :rtype: Program
+
+        """
+
+        for action in self.actions:
+            assert action[0] == 0, "Program must be valid Protoquil"
+            gate = action[1]
+            assert not isinstance(gate, Measurement), "Program cannot contain measurements"
+            assert not isinstance(gate, While) and not isinstance(gate, If), \
+                "Program cannot contain control flow"
+
+        daggered = Program()
+
+        for gate in self.defined_gates:
```

Comment:
Is there a way to have the user specify if they define the inverse of a defgate?  one way would be to pass a map between defgate names such that the key value pair represents the defgate and inv_defgate.  @tarballs-are-good brought up a good point that we want to give the user the option to predefine their inverse defgate so we avoid defgating all over the place.  We might want to have the following:
`inv_list`  which maps a gate_name to it's inverse, `suffix` defined initially as `INV` if we want.  Furthermore, we might consider adding a field ('dagger') to a `Program` object representing the daggered program.  One might need this because the inverse program might be difficult to implement.